### PR TITLE
Convert to Typescript

### PR DESCRIPTION
**Pull request by declanslevin**

### CODE REVIEW

File: lib/game.test.ts [hunk 1]
```diff
@@ -1,9 +1,59 @@
...
+        break;
+      case "Human":
+        players = [new HumanPlayer("Player 1"), new HumanPlayer("Player 2")];
+        game = new Game(players, ...gameArgs);
+        regularisePlayers(game);
+        break;
+    }
+  } else {
+    players = [new HumanPlayer(), new CpuPlayer()];
+    game = new Game(players, ...gameArgs);
```
Comment (alsuren) [inline lib/game.test.ts]: either throw new error here, or also call regularisePlayers()

File: lib/game.test.ts [hunk 2]
```diff
@@ -1,9 +1,59 @@
...
+  ] = [grid, nextPlayer, choices, winner, mode];
+  let game: Game;
+  let players: [Player, Player];
+
+  if (vs) {
+    switch (vs) {
+      case "Cpu":
+        players = [new HumanPlayer("You"), new CpuPlayer("CPU")];
+        game = new Game(players, ...gameArgs);
+        regularisePlayers(game);
```
Comment (alsuren) [inline lib/game.test.ts]: regularisePlayers(game, players);

=>

game.players.X = players[0]
game.players.Y = players[1]

File: lib/checkWin.ts [hunk 3]
```diff
@@ -41,16 +45,24 @@ const checkForDraw = async (game) => {
...
+  game: Game,
+  win: boolean,
+  draw: boolean
+): Promise<void> => {
   if (win || draw) {
-    let log;
+    let log: string;
     let gameOverObj;
     if (win) {
+      if (!game.winner) {
```
Comment (declanslevin) [inline lib/checkWin.ts:62]: - Set game.winnner to be of type Team | null
- Assign game.winner to a variable to help with type narrowing

File: lib/checkWin.ts [hunk 4]
```diff
@@ -64,29 +76,32 @@ const logGameOver = async (game, win, draw) => {
...
-exports.checkForGameOver = checkForGameOver;
-exports.waitForGameOver = waitForGameOver;
+export {
+  mapGridValues,
+  checkMappedGridForWin,
+  checkForDraw,
+  checkForWin,
+  checkForGameOver,
+  waitForGameOver,
+};
```
Comment (declanslevin) [inline lib/checkWin.ts:108]: Replace this export section with export const on the functions themselves (do this for all files)

File: lib/game.ts [hunk 5]
```diff
@@ -1,6 +1,27 @@
...
+
+interface Players {
+  O: Player;
+  X: Player;
+}
+
+export type Team = "O" | "X";
+
+export class Game {
+  // TODO
```
Comment (declanslevin) [inline lib/game.ts]: Remove

File: lib/game.ts [hunk 6]
```diff
@@ -1,6 +1,27 @@
...
+  grid: Grid;
+  players: Players;
+  nextPlayer: Team;
+  choices: number[];
+  winner?: string | null;
+  mode?: string | null;
```
Comment (declanslevin) [inline lib/game.ts]: Remove this as it's never actually set

File: lib/game.ts [hunk 7]
```diff
@@ -1,6 +1,27 @@
...
+
+export type Team = "O" | "X";
+
+export class Game {
+  // TODO
+  grid: Grid;
+  players: Players;
+  nextPlayer: Team;
+  choices: number[];
+  winner?: string | null;
```
Comment (declanslevin) [inline lib/game.ts:19]: Change this to Team | null

File: lib/game.ts [hunk 8]
```diff
@@ -10,39 +31,42 @@ class Game {
...
   }
-  randomTeams() {
+  randomPlayersObject(players: [Player, Player]): Players {
+    const swap = Math.random() < 0.5;
+    if (swap) {
+      return {
+        X: players[0],
+        O: players[1],
+      };
+    }
```
Comment (declanslevin) [inline lib/game.ts]: else {

File: lib/game.ts [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    return {
+      O: players[0],
+      X: players[1],
+    };
+  }
+  randomTeams(): Team[] {
```
Comment (declanslevin) [inline lib/game.ts]: Remove all dead code

File: lib/game.ts [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+  }
+  logGrid(): void {
+    this.players.O.logGrid(this);
+    this.players.X.logGrid(this);
+  }
+  log(message: string): void {
+    this.players.O.log(message);
+    this.players.X.log(message);
+  }
+  send(messageObj: any): void {
```
Comment (declanslevin) [inline lib/game.ts:96]: Check for "any" types and assign types (backlog this until TS in frontend)

File: lib/play.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  await playerTurn(game, player);
+  game.setNextPlayer();
+};
+
+const receiveGameModeChoice = async (player: HumanPlayer): Promise<string> => {
+  return new Promise((resolve) => {
+    if (player.ws) {
+      player.ws.on("message", (message: string) => {
+        let msg = JSON.parse(message);
+        // TODO: Message validation
```
Comment (declanslevin) [inline lib/play.ts:24]: Should be covered by the "check for any type" issue (to be created)

File: lib/play.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+      });
+    }
+  });
+};
+
+const play = async (game: Game, players?: Player[]): Promise<void> => {
+  if (players) {
+    game.sendInitialBrowserState();
+    // game.setPlayers(players);
+    // game.setPlayOrder();
```
Comment (declanslevin) [inline lib/play.ts]: remove commented code

File: lib/play.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  } else {
+    await waitForGameOver(game);
+  }
+};
+
+const runPlayLoop = async (
+  lobby: Lobby,
+  player: HumanPlayer
+): Promise<void> => {
+  console.log("Starting runPlayLoop");
```
Comment (declanslevin) [inline lib/play.ts]: remove logging

File: lib/play.ts [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+        lobby.addGame(game);
+        await play(game, players);
+      } else {
+        const game = lobby.getGameFromPlayer(player);
+        await play(game);
+      }
+    }
+    playAgainResult = await playAgain(player);
+    if (!playAgainResult) {
+      player.log("Thank you for playing!");
```
Comment (declanslevin) [inline lib/play.ts]: Either set while loop to while(true) or move the player.log outside the while loop and delete this if statement

File: lib/playAgain.ts [hunk 15]
```diff
@@ -9,7 +9,7 @@ const recievePlayAgainChoice = (player) => {
   });
 };
 
-const isValidPlayAgainAnswer = (player, answer) => {
+const isValidPlayAgainAnswer = (player: any, answer: any) => {
```
Comment (declanslevin) [inline lib/playAgain.ts]: set player: Player
give answer a type 
no return type

File: lib/playAgain.ts [hunk 16]
```diff
@@ -9,7 +9,7 @@ const recievePlayAgainChoice = (player) => {
   });
 };
 
-const isValidPlayAgainAnswer = (player, answer) => {
```
Comment (declanslevin) [inline lib/playAgain.ts:12]: https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

Comment (declanslevin) [inline lib/playAgain.ts:12]: Possibly bin altogether as this is a hangover from readline days

File: lib/playAgain.ts [hunk 17]
```diff
@@ -18,7 +18,7 @@ const isValidPlayAgainAnswer = (player, answer) => {
   }
 };
 
-const playAgain = async (player) => {
+const playAgain = async (player: any): Promise<boolean> => {
```
Comment (declanslevin) [inline lib/playAgain.ts]: set player type

File: lib/playAgain.ts [hunk 18]
```diff
@@ -18,7 +18,7 @@ const isValidPlayAgainAnswer = (player, answer) => {
...
   while (true) {
     let answer = await recievePlayAgainChoice(player);
```
Comment (declanslevin) [inline lib/playAgain.ts:27]: answer: string

File: lib/player.ts [hunk 19]
```diff
@@ -52,12 +67,13 @@ class Player {
       console.log(message);
     }
   }
-  send(messageObj) {
+  // TODO: Type this properly
```
Comment (declanslevin) [inline lib/player.ts:70]: Check issue for setting message types

**Review by alsuren — APPROVED** [review]
Looks good. Backlog the things that aren't trivial and ship it.

File: lib/game.ts (hunk 7 above)
Comment (declanslevin) [reply lib/game.ts:19]: Logged under #32

File: lib/game.ts (hunk 10 above)
Comment (declanslevin) [reply lib/game.ts:96]: Logged under #33

File: lib/play.ts (hunk 11 above)
Comment (declanslevin) [reply lib/play.ts:24]: See #33

File: lib/playAgain.ts (hunk 16 above)
Comment (declanslevin) [reply lib/playAgain.ts:12]: #34

Comment (declanslevin) [reply lib/playAgain.ts:12]: #34

File: lib/player.ts (hunk 19 above)
Comment (declanslevin) [reply lib/player.ts:70]: #33

File: lib/checkWin.ts (hunk 3 above)
Comment (declanslevin) [reply lib/checkWin.ts:62]: #32

File: lib/playAgain.ts (hunk 15 above)
Comment (declanslevin) [reply lib/playAgain.ts]: #34

File: lib/checkWin.ts (hunk 4 above)
Comment (declanslevin) [reply lib/checkWin.ts:108]: #35

**Pull request merged**